packages/dashboard-frontend/src/services/workspace-client/devworkspace/DevWorkspaceDefaultPluginsHandler.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2018-2021 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+import { injectable } from 'inversify';
+import { api } from '@eclipse-che/common';
+import * as ServerConfigApi from '../../dashboard-backend-client/serverConfigApi';
+import { createHash } from 'crypto';
+import devfileApi from '../../devfileApi';
+import { V1alpha2DevWorkspaceSpecTemplateComponents } from '@devfile/api';
+import * as DwApi from '../../dashboard-backend-client/devWorkspaceApi';
+
+const DEFAULT_PLUGIN_ATTRIBUTE = 'default-plugin';
+
+/**
+ * This class manages the default plugins defined in
+ * the DevWorkspace's spec.template.components array.
+ */
+@injectable()
+export class DevWorkspaceDefaultPluginsHandler {
+  public async handle(workspace: devfileApi.DevWorkspace, editor: string): Promise<void> {
+    const defaultPlugins = await ServerConfigApi.getDefaultPlugins(editor);
+    this.handleUriPlugins(workspace, defaultPlugins);
+    this.patchWorkspaceComponents(workspace);
+  }
+
+  private async handleUriPlugins(workspace: devfileApi.DevWorkspace, defaultPlugins: string[]) {
+    const defaultUriPlugins = new Set(defaultPlugins.filter(this.isUri));
+
+    this.removeOldDefaultUriPlugins(workspace, defaultUriPlugins);
+
+    defaultUriPlugins.forEach(plugin => {
+      const name = createHash('MD5').update(plugin).digest('hex').substring(0, 20).toLowerCase();
+      this.addDefaultPluginByUri(workspace, 'default-' + name, plugin);
+    });
+  }
+
+  private isUri(str: string) {
```

Comment:
my understanding is that if the definition of the plugin is not valid (not URI), there will be no error message, and the plugins will not be processed at all. Probably a good idea, but can we possibly at least log it somehow

Reply:
Yes, that's how it works, and sounds good, I've updated the PR with logging:
https://github.com/eclipse-che/che-dashboard/blob/6aeacf6c1c0eb8f683b72e6dacd63b27aedc7974/packages/dashboard-frontend/src/services/workspace-client/devworkspace/DevWorkspaceDefaultPluginsHandler.ts#L40